source/braille.py

```diff
@@ -1444,11 +1445,141 @@ def _setCursor(self, info: textInfos.TextInfo):
+		"""
+		self._readingUnitContainsSelectedCharacters = False
+		if not config.conf["braille"]["showSelection"]:
+			self._currentSelection = None
+			return api.getReviewPosition().copy()
+		info: textInfos.TextInfo
+		try:
+			info = self.obj.makeTextInfo(textInfos.POSITION_SELECTION)
+		except (LookupError, RuntimeError):
+			self._currentSelection = None
+			return self._collapsedReviewPosition()
+		if info.isCollapsed:
+			# Cursor
+			self._currentSelection = None
+			return self._collapsedReviewPosition()
+		# Selection
+		if self._currentSelection is None or self._currentSelection != info:
+			# Selection changed, update also review position
+			self._currentSelection = info.copy()
+			if self.obj.isTextSelectionAnchoredAtStart:
+				self._currentSelection.move(textInfos.UNIT_CHARACTER, -1, "end")
+			self._currentSelection.collapse(end=self.obj.isTextSelectionAnchoredAtStart)
+			# Enqueue to avoid recursion error
+			queueHandler.queueFunction(queueHandler.eventQueue, api.setReviewPosition, self._currentSelection)
+			self._currentSelection = info.copy()
+			return info
+		# Selection unchanged
+		readingUnit: textInfos.TextInfo = api.getReviewPosition().copy()
+		readingUnit.expand(self._getReadingUnit())
+		if (
+			readingUnit.start >= info.end
+			or readingUnit.end <= info.start
+		):
+			# Reading unit containing review position is outside of selection
+			return self._collapsedReviewPosition()
+		else:
+			# Reading unit contains selected characters
+			# All characters are not necessarily selected
+			if readingUnit.start < self._currentSelection.start:
+				readingUnit.start = self._currentSelection.start
+			if readingUnit.end > self._currentSelection.end:
+				readingUnit.end = self._currentSelection.end
+			self._readingUnitContainsSelectedCharacters = True
+			return readingUnit
+
+	def _collapsedReviewPosition(self) -> textInfos.TextInfo:
+		"""Gets collapsed review position.
+		:return: collapsed review position
+		"""
+		info: textInfos.TextInfo = api.getReviewPosition().copy()
+		# Info should be collapsed, but it is not always, at least when
+		# switching from focus mode to browse mode.
+		info.collapse()
+		return info
 
-	def _getSelection(self):
-		return api.getReviewPosition().copy()
+	def update(self):
+		"""Updates this region.
+		When "show selection" is enabled and reading unit contains selected
+		characters, adds functionality to properly move to the next and previous line,
+		to start and end of current control, to start and end of current line,
+		to scroll forward and back to the next and previous line,
+		and to move to focus.
+		"""
+		previousReadingUnit: textInfos.TextInfo | None = getattr(self, "_readingInfo", None)
+		super().update()
+		if (
+			not config.conf["braille"]["showSelection"]
+			or not self._readingUnitContainsSelectedCharacters
+			or self._currentScript is None
+			or previousReadingUnit is None
+		):
+			return
+		from globalCommands import GlobalCommands

```

Comment:
I'm really concerned about the code below, specifically the way `TextInfoRegion` becomes dependent on GlobalCommands scripts here. @seanbudd, could you please consider this in your review. I'd also really appreciate @michaelDCurran's opinion here.

Reply:
I think that at least this approach it is important to know which script is currently running so that moving from position to position within selection would work with braille. If no script is running nothing should happen. If some other script is running, do not scroll braille because it would likely require handling more cases. So this solution does not try to handle all cases but those I have thought are important. Although all scripts which normally scroll braille, do not scroll braille in this implementation, when braille is displaying something within selection (or reading unit which contains selected characters), scripts are running otherwise as they normally run. This pr should not block any scripts.

Reply:
> I think that at least this approach it is important to know which script is currently running so that moving from position to position within selection would work with braille. 

But why does this code need to know what exact script is running? If behavior should be different for some scripts, it should be the script to dictate the behavior, not the region changing its behavior based on the script that triggered the update. Ideally, the script should tell the region how to change, not vise versa.

In your code:

1. script_braille_nextLine is executed, which calls `nextLine` on the region
2. `nextLine` on the region calls `_setCursor`
3. `_setCursor` calls `api.setReviewPosition`
4. `api.setReviewPosition` calls `braille.handler.handleReviewMove`
5. `handleReviewMove` queues a caret update (you're setting the script there)to update)

I assume it should be possible to recalculate the selection in `nextLine`, but i'm not even convinced that it should be that complex. As stated over and over again, this code concerns me in a major way. Its complexity makes it almost impossible to track down what's actually happening, what's necessary and what not.
Again a call for @michaelDCurran to jump on the train here.

Reply:
When this pr code is executed there is no script running anymore because this code (expect getting running script in `handleReviewMove`) is executed in `pumpAll` which is run at end of core cycle.

When `ReviewTextInfoRegion.update` is executed and if review position is within selection (assuming there is selected text at this moment), and when given script has run, it has to be known what part of current reading unit should be displayed in braille. To part of reading unit to be displayed depends on which script was executed. For example: if script was scrollBack, it means that end of reading unit should be displayed. If script was previous or next line or scroll forward, then start of reading unit should be displayed.

If moving within current reading unit with scrollForward and scrollBack, this code should never be executed.

I suppose that if scripts should instruct region, then they should be aware about selection, if show selection is enabled, and if review position is within selection (reading unit contains selected characters).